File: iOS/Schema/Sources/Widgets/Button/Button.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

public struct Button: RawWidget, ClickedOnComponent, AutoInitiableAndDecodable {
    
    // MARK: - Public Properties
    public let text: Expression<String>
    public let styleId: String?
    public let onPress: [RawAction]?
    public var clickAnalyticsEvent: AnalyticsClick?
    public var widgetProperties: WidgetProperties

// sourcery:inline:auto:Button.Init
    public init(
        text: Expression<String>,
        styleId: String? = nil,
        onPress: [RawAction]? = nil,
        clickAnalyticsEvent: AnalyticsClick? = nil,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.text = text
        self.styleId = styleId
        self.onPress = onPress
        self.clickAnalyticsEvent = clickAnalyticsEvent
        self.widgetProperties = widgetProperties
    }
// sourcery:end
}

File: schema/schema-ruby-erb/model_generation/Models/Widgets/button.rb
#   Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA

#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
 
#       http://www.apache.org/licenses/LICENSE-2.0
 
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

require_relative '../../Synthax/Attributes/variable.rb'
require_relative '../../Synthax/Attributes/list.rb'
require_relative '../../Synthax/Types/built_in_type.rb'
require_relative '../base_component.rb'
require_relative '../Analytics/analytics_events.rb'
require_relative '../Analytics/analytics_models.rb'
require_relative '../Action/action.rb'
require_relative 'widget.rb'

class Button < BaseComponent

    def initialize
        variables = [
            Variable.new(:name => "text", :typeName => BasicTypeKeys.string, :isBindable => true),
            Variable.new(:name => "styleId", :typeName => BasicTypeKeys.string, :isOptional => true),
            List.new(:name => "onPress", :typeName => Action.new.name, :isOptional => true), #todo create action
            Variable.new(:name => "clickAnalyticsEvent", :typeName => AnalyticsClick.new.name, :isOptional => true)
        ]
        synthax_type = BuiltInType.new(
            :name => self.name,
            :variables => variables,
            :package => "br.com.zup.beagle.widget.ui",
            :inheritFrom => [
               Widget.new.name,
               "ClickedOnComponent" #todo create
            ]
        )

        super(synthax_type)

    end
    
end